stubs/SQLAlchemy/sqlalchemy/engine/url.pyi

```diff
@@ -1,46 +1,60 @@
-from typing import Any
+from _typeshed import Self, SupportsItems
+from collections.abc import Iterable, Mapping, Sequence
+from typing import Any, NamedTuple
 
-from ..util import memoized_property
+from ..util import immutabledict, memoized_property
+from .interfaces import Dialect
 
-class URL:
-    def __new__(cls, *arg, **kw): ...
+# stub-only helper class
+class _URLTuple(NamedTuple):
+    drivername: str
+    username: str | None
+    password: str | object | None  # object that produces a password when called with str()
+    host: str | None
+    port: int | None
+    database: str | None
+    query: immutabledict[str, str | tuple[str, ...]]
+
+_Query = Mapping[str, str | Sequence[str]] | Sequence[tuple[str, str | Sequence[str]]]
+
+class URL(_URLTuple):
     @classmethod
     def create(
         cls,
-        drivername,
-        username: Any | None = ...,
-        password: Any | None = ...,
-        host: Any | None = ...,
-        port: Any | None = ...,
-        database: Any | None = ...,
-        query=...,
-    ): ...
+        drivername: str,
+        username: str | None = ...,
+        password: str | object | None = ...,  # object that produces a password when called with str()
+        host: str | None = ...,
+        port: int | None = ...,
+        database: str | None = ...,
+        query: _Query | None = ...,
+    ) -> URL: ...
     def set(
-        self,
-        drivername: Any | None = ...,
-        username: Any | None = ...,
-        password: Any | None = ...,
-        host: Any | None = ...,
-        port: Any | None = ...,
-        database: Any | None = ...,
-        query: Any | None = ...,
-    ): ...
-    def update_query_string(self, query_string, append: bool = ...): ...
-    def update_query_pairs(self, key_value_pairs, append: bool = ...): ...
-    def update_query_dict(self, query_parameters, append: bool = ...): ...
-    def difference_update_query(self, names): ...
+        self: Self,
+        drivername: str | None = ...,
+        username: str | None = ...,
+        password: str | object | None = ...,
+        host: str | None = ...,
+        port: int | None = ...,
+        database: str | None = ...,
+        query: _Query | None = ...,
+    ) -> Self: ...
+    def update_query_string(self: Self, query_string: str, append: bool = ...) -> Self: ...
+    def update_query_pairs(self: Self, key_value_pairs: Iterable[tuple[str, str]], append: bool = ...) -> Self: ...
+    def update_query_dict(self: Self, query_parameters: SupportsItems[str, str | Sequence[str]], append: bool = ...) -> Self: ...
+    def difference_update_query(self, names: Iterable[str]) -> URL: ...
     @memoized_property
```

Comment:
A plain `@property` might be better, because type checkers special-case `@property`.

Reply:
As far as I know that special-casing isn't necessary in modern type checkers. This prevents stubtest problem and is closer to the actual implementation, which is usually a plus.

Reply:
What is a modern type checker, if mypy and pyright both special-case property?

Reply:
Both don't need to special-case it, as far as I know. There were discussions about removing the special casing.

Reply:
But non-special-cased `@property` is really quite terrible: #5987 (note that this isn't merged)